File: 1-树/4-伸展树Splay/zentorwie/SplayTree.c
//
//  SplayTree.c
//  Created by zentorwie on 2014-03-17

#include <stdio.h>
#include <stdlib.h>

#define KEY_TYPE int

struct SplayNode {
  KEY_TYPE Key;
  struct SplayNode* Left;
  struct SplayNode* Right;
};

typedef struct SplayNode* SplayTree;
SplayTree Initialize(void);
SplayTree Insert(KEY_TYPE x, SplayTree T);
SplayTree Remove(KEY_TYPE x, SplayTree T);
SplayTree Splay(int x, SplayTree Root);
SplayTree Find(KEY_TYPE x, SplayTree T);
SplayTree FindMin(SplayTree T);
SplayTree FindMax(SplayTree T);
SplayTree RotateWithLeft(SplayTree T);
SplayTree RotateWithRight(SplayTree T);
void InOrder(SplayTree T);
static KEY_TYPE Retrieve(SplayTree Node);

static SplayTree NullNode = NULL;

int main()
{
  SplayTree T = Initialize();

  T = Insert(1, T);

  T = Insert(2, T);
  T = Insert(3, T);
  T = Insert(4, T);
  T = Insert(9, T);
  T = Insert(8, T);

  InOrder(T);
  puts("haha");
  T = Find(4, T);

  InOrder(T);
  return 0;
}

SplayTree Initialize(void) 
{
  if (NullNode == NULL) {
    NullNode = (SplayTree)malloc(sizeof(struct SplayNode));
    if (NullNode == NULL) {
      printf("Out of space!\n");
      exit(1);
    }
    NullNode->Left = NullNode->Right = NullNode;
  }
  return NullNode;
}

SplayTree Insert(KEY_TYPE x, SplayTree T)
{
  static SplayTree NewNode = NULL;
  
  if (NewNode == NULL) {
    NewNode = (SplayTree)malloc(sizeof(struct SplayNode));
    if (NewNode == NULL) {
      printf("Out of space!\n");
      exit(1);
    }
  }
  T->Key = x;
  
  if (T == NullNode) {
    NewNode->Left = NewNode->Right = NullNode;
    T = NewNode;
  }
  else {
    T = Splay(x, T);
    if (x < T->Key) {
      NewNode->Left = T->Left;
      NewNode->Right = T;
      T->Left = NullNode;
      T = NewNode;
    }
    else if (x > T->Key) {
      NewNode->Right = T->Right;
      NewNode->Left = T;
      T->Right = NullNode;
      T = NewNode;
    }
    else 
      return T;
  }
  
  NewNode = NULL;
  return T;
}


SplayTree Remove(KEY_TYPE x, SplayTree T)
{
  SplayTree NewTree;

  if (T != NullNode) {
    T = Splay(x, T);
    if (x == T->Key) {
      if (T->Left == NullNode) 
	NewTree = T->Right;
      else {
	NewTree = T->Left;
	NewTree = Splay(x, NewTree);
	NewTree->Right = T->Right;
      }
      free(T);
      T = NewTree;
    }
  }
  return T;
}

SplayTree Splay(KEY_TYPE x, SplayTree Root)
{
  static struct SplayNode Header;
  SplayTree LeftTreeMax, RightTreeMin;

  Header.Left = Header.Right = NullNode;
  LeftTreeMax = RightTreeMin = &Header;
  NullNode->Key = x;
  
  while(x != Root->Key) {
    if (x < Root->Key) {
      if (x < Root->Left->Key)
	Root = RotateWithLeft(Root);
      if (Root->Left == NullNode) 
	break;
      RightTreeMin->Left = Root;
      RightTreeMin = Root;
      Root = Root->Left;
    }
    else {
      if (x > Root->Key) 
	Root = RotateWithRight(Root);
      if (Root->Right == NullNode)
	break;
      LeftTreeMax->Right = Root;
      LeftTreeMax = Root;
      Root = Root->Right;
    }
  } /* while (x != Root->Key) */

  /* Reassemble */
  LeftTreeMax->Right = Root->Left;
  RightTreeMin->Left = Root->Right;
  Root->Left = Header.Right;
  Root->Right = Header.Left;

  return Root;
}

SplayTree Find(KEY_TYPE X, SplayTree T)
{
  SplayTree Target;
  Target = Splay(X, T);
  if (T == NULL || Target->Key != X) {
    printf("Not found!\n");
    return NULL;
  }
  else
    return T;
}

SplayTree FindMin(SplayTree T)
{
  if (T == NULL)
    return NULL; // Not found.
  if (T->Left == NULL)
    return T;
  else
    return FindMin(T->Left);
}

SplayTree FindMax(SplayTree T)
{
  if (T == NULL)
    return NULL; // Not found
  if (T->Right == NULL)
    return T;
  else
    return FindMax(T->Right);
}

SplayTree RotateWithLeft(SplayTree T)
{
  SplayTree NewRoot;

  NewRoot = T->Left;
  T->Left = NewRoot->Right;
  NewRoot->Right = T;

  return NewRoot;
}

SplayTree RotateWithRight(SplayTree T)
{
  SplayTree NewRoot;
  NewRoot = T->Right;
  T->Right = NewRoot->Left;
  NewRoot->Left = T;

  return NewRoot;
}


void InOrder(SplayTree T)
{
  if (T == NullNode)
    return;
  InOrder(T->Left);
  printf("%d\n", T->Key);
  InOrder(T->Right);
}

static KEY_TYPE Retrieve(SplayTree Node)
{
  return Node->Key;
}

File: 1-树/4-伸展树Splay/caven/SplayTree.java
    package com.kiritor;  
      
    /**伸展树 
     * @author caven*/  
    public class SplayTree {  
      
        static class BinaryNode {  
            // Constructors  
            BinaryNode(Comparable theElement) {  
                this(theElement, null, null);  
            }  
      
            BinaryNode(Comparable theElement, BinaryNode lt, BinaryNode rt) {  
                element = theElement;  
                left = lt;  
                right = rt;  
            }  
            Comparable element;  
            BinaryNode left;   
            BinaryNode right;  
        }  
      
        private BinaryNode root;  
        private static BinaryNode nullNode;  
        static   
        {  
            nullNode = new BinaryNode(null);  
            nullNode.left = nullNode.right = nullNode;  
        }  
      
        private static BinaryNode newNode = null; //用于插入的操作  
        private static BinaryNode header = new BinaryNode(null);//用于调整操作   
          
        public SplayTree() {  
            root = nullNode;  
        }  
        public void insert(Comparable x) {  
            if (newNode == null)  
                newNode = new BinaryNode(x);//新建一个结点  
            //根结点为空则新建的结点作为根结点  
            if (root == nullNode) {  
                newNode.left = newNode.right = nullNode;  
                root = newNode;  
            } else {  
                root = splay(x, root);//调整  
                if (x.compareTo(root.element) < 0) {  
                    newNode.left = root.left;  
                    newNode.right = root;  
                    root.left = nullNode;  
                    root = newNode;  
                } else if (x.compareTo(root.element) > 0) {  
                    newNode.right = root.right;  
                    newNode.left = root;  
                    root.right = nullNode;  
                    root = newNode;  
                } else  
                    return;  
            }  
            newNode = null;   
        }  
      
        public void remove(Comparable x) {  
            BinaryNode newTree;  
            root = splay(x, root);  
            if (root.element.compareTo(x) != 0)  
                return; // Item not found; do nothing  
            if (root.left == nullNode)  
                newTree = root.right;  
            else {  
                newTree = root.left;  
                newTree = splay(x, newTree);  
                newTree.right = root.right;  
            }  
            root = newTree;  
        }  
        public Comparable findMin() {  
            if (isEmpty())  
                return null;  
            BinaryNode ptr = root;  
            while (ptr.left != nullNode)  
                ptr = ptr.left;  
            root = splay(ptr.element, root);  
            return ptr.element;  
        }  
      
        public Comparable findMax() {  
            if (isEmpty())  
                return null;  
            BinaryNode ptr = root;  
            while (ptr.right != nullNode)  
                ptr = ptr.right;  
            root = splay(ptr.element, root);  
            return ptr.element;  
        }  
        public Comparable find(Comparable x) {  
            root = splay(x, root);  
            if (root.element.compareTo(x) != 0)  
                return null;  
            return root.element;  
        }  
        public void makeEmpty() {  
            root = nullNode;  
        }  
      
        public boolean isEmpty() {  
            return root == nullNode;  
        }  
        public void printTree() {  
            if (isEmpty())  
                System.out.print("Empty tree  ");  
            else  
                printTree(root);  
        }  
      
        private BinaryNode splay(Comparable x, BinaryNode t) {  
            BinaryNode leftTreeMax, rightTreeMin;  
            header.left = header.right = nullNode;  
            leftTreeMax = rightTreeMin = header;  
            nullNode.element = x;   
            for (;;)  
                if (x.compareTo(t.element) < 0) {  
                    if (x.compareTo(t.left.element) < 0)  
                        t = rotateWithLeftChild(t);  
                    if (t.left == nullNode)  
                        break;  
                    rightTreeMin.left = t;  
                    rightTreeMin = t;  
                    t = t.left;  
                } else if (x.compareTo(t.element) > 0) {  
                    if (x.compareTo(t.right.element) > 0)  
                        t = rotateWithRightChild(t);  
                    if (t.right == nullNode)  
                        break;  
                    // Link Left  
                    leftTreeMax.right = t;  
                    leftTreeMax = t;  
                    t = t.right;  
                } else  
                    break;  
      
            leftTreeMax.right = t.left;  
            rightTreeMin.left = t.right;  
            t.left = header.right;  
            t.right = header.left;  
            return t;  
        }  
      
          
        static BinaryNode rotateWithLeftChild(BinaryNode k2) {  
            BinaryNode k1 = k2.left;  
            k2.left = k1.right;  
            k1.right = k2;  
            return k1;  
        }  
      
          
        static BinaryNode rotateWithRightChild(BinaryNode k1) {  
            BinaryNode k2 = k1.right;  
            k1.right = k2.left;  
            k2.left = k1;  
            return k2;  
        }  
      
        private void printTree(BinaryNode t) {  
            if (t != t.left) {  
                printTree(t.left);  
                System.out.print(t.element.toString()+"  ");  
                printTree(t.right);  
            }  
        }  
      
        public static void main(String[] args) {  
            SplayTree tree = new SplayTree();  
            tree.insert(12);  
            tree.insert(8);  
            tree.insert(2);  
            tree.insert(4);  
            tree.insert(14);  
            tree.insert(16);  
            tree.insert(6);  
            tree.insert(1);  
            tree.insert(11);  
            tree.remove(8);   
            System.out.println("被查找的节点:" + tree.find(11));  
            System.out.println("此时的根:" + tree.root.element);  
            System.out.println("被查找的节点:" + tree.find(12));  
            System.out.println("此时的根:" + tree.root.element);  
            System.out.println("被查找的节点:" + tree.find(11));  
            System.out.println("此时的根:" + tree.root.element);  
            System.out.println("伸展树值情况:");  
            tree.printTree();   
      
        }  
    }  